### PR TITLE
Stable Config file: target system properties in process_arguments and support template variables in YamlParser

### PR DESCRIPTION
**Pull request by mtoffl01**
# What Does This Do
This PR adds support for matching system property keys in the Stable Configuration selector, even when those properties include values. For example, a selector with `-Darg1` will now correctly match a Java process started with `-Darg1=value`.

It also introduces support for template variables in the Stable Configuration file schema. These variables can resolve values from process arguments and environment variables. (See [reference](https://docs.google.com/document/d/1MNI5d3g6R8uU3FEWf2e08aAsFcJDVhweCPMjQatEb0o/edit?tab=t.hpao9cjxqjyn#heading=h.66zxnu558h5b) for more details.)

## Notes & Limitations:
**Scope Limitation**: This change restricts `process_arguments` selectors and template variables to System Properties only.
**Required Format**: System Properties must be referenced using their full -D prefix, without specifying a value (e.g., use `-Dkey`, not `key` nor `-Dkey=value`).

# Motivation
Stable Config users must be able to target services based on the presence of a specific System Property, regardless of its value. For instance, `-Darg1` should match any process started with `-Darg1=abcd`, `-Darg1=xyz`, etc.

Currently, our only known use case for `process_arguments` selectors involves System Properties. If a need arises to support additional types of JVM arguments in the future, we can revisit and expand this functionality.

# Additional Notes

# Contributor Checklist

- Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any usefull labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentBootstrap.java [hunk 1]
```diff
@@ -381,8 +382,12 @@ private static List<File> getAgentFilesFromVMArguments() {
       // - On IBM-based JDKs since at least 1.7
       // This prevents custom log managers from working correctly
       // Use reflection to bypass the loading of the class~
-      for (final String argument : CLIHelper.getVmArgs()) {
+      Map<String, String> vmArgs = CLIHelper.getVmArgs();
+      for (final Map.Entry<String, String> entry : vmArgs.entrySet()) {
+        final String argument = entry.getKey();
         if (argument.startsWith(JAVA_AGENT_ARGUMENT)) {
+          // TODO: Modify CLIHelper to parse key-vals on JAVA_AGENT_ARGUMENT arguments, so we don't
+          // have to do the parsing here.
```
Comment (mtoffl01) [inline dd-java-agent/src/main/java/datadog/trace/bootstrap/AgentBootstrap.java:390]: Nothing is stopping me from making this change in CLIHelper.parseVmArgs now; the only reason I did not go ahead and do it is to avoid introducing unnecessary complexity / expanding the changes in this PR outside of its intended scope.

File: components/yaml/src/test/groovy/datadog/yaml/YamlParserTest.groovy [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package datadog.yaml
+
+import datadog.trace.test.util.DDSpecification
+
+class YamlParserTest extends DDSpecification{
```
Comment (mtoffl01) [inline components/yaml/src/test/groovy/datadog/yaml/YamlParserTest.groovy:5]: These tests are in progress. I want to use helper methods from another test class, but blocked on how best to do this: [[slack link](https://dd.slack.com/archives/C047YAMGZ7T/p1744906101173789)]

> I wrote [helper functions inside a test class](https://github.com/DataDog/dd-trace-java/blob/master/internal-api/src/test/groovy/datadog/trace/bootstrap/config/provider/StableConfigSourceTest.groovy#L168-L188) on the internal-api path that I'd like to use in other test classes, at other paths. These methods are already used in another test file at the same path ([example](https://github.com/DataDog/dd-trace-java/blob/master/internal-api/src/test/groovy/datadog/trace/bootstrap/config/provider/StableConfigParserTest.groovy#L45)), but now I'd like to use those same helper functions inside the components path.
Where is the best place to store these helper functions? Is there some kind of global test-util directory for cases like this? Or, would I maybe put them on the DDSpecification class?

**Review by PerfectSlayer — COMMENTED** [review]
Hey @mtoffl01 

I feel there is a mix of concerns about how the `CLIHelper` changed.
As it’s a common composent, we would expect him to keep providing the JVM arguments. (`-Dkey=value` being a process argument)

While I think that parsing JVM properties is interesting (and needed), it should not replace the process argument API. So what about having:
* one way to retrieve arguments as a list of string (as before)
* one way of querying JVM properties (get the key value if any)
* but still do the parsing in one pass

Overall, I also think the common component needs:
* A complete Javadoc
* Tests in order to check edge case behavior (like property redefinition, empty property value, etc...).
What do you thing about it?

And my mid-term goal is to make this component evolves to a `system` / `env` component that will be provide any kind of information about the execution context (JDK version, vendor, native-image, etc...).

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 3]
```diff
@@ -9,25 +9,30 @@
...
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 
+/** Helper class for retrieving and parsing JVM arguments. */
 public final class CLIHelper {
-  private static final List<String> VM_ARGS = findVmArgs();
+  private static final Map<String, String> VM_ARGS = findVmArgs();
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:19]: I'd prefer to keep this as a list, and instead provide a second method to query the arguments as a map.

This second method could lazily parse the list of arguments and cache the results.

That way we continue to have the simple approach for the common case, while still allowing querying as a map.

Note you don't have to use `Map` as the return value of this second method - you could just have a `getVmArg(String)` method which returned a string, underneath this could trigger lazy parsing.

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 4]
```diff
@@ -9,18 +9,102 @@
...
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
 
+/** Helper class for retrieving and parsing JVM arguments. */
 public final class CLIHelper {
   private static final List<String> VM_ARGS = findVmArgs();
+  private static Map<String, String> VM_ARGS_MAP = Collections.emptyMap();
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:20]: I'd initialize this as `new HashMap<>()` and make it final - an empty hash map doesn't take up that much space compared to all the other caches needed for tracing, and it would simplify things (and avoid a concurrency issue - see below)

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 5]
```diff
@@ -9,18 +9,102 @@
...
 
   public static List<String> getVmArgs() {
     return VM_ARGS;
   }
 
+  public static String getArgValue(String key) {
+    int numArgs = VM_ARGS.size();
+    // Lazy population of cache
+    synchronized (VM_ARGS_MAP) {
+      // Double check after acquiring lock
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:31]: note this is not technically a 'double-check' as we didn't do anything before acquiring the lock

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 6]
```diff
@@ -9,18 +9,102 @@
...
+      if (VM_ARGS_MAP.containsKey(key)) {
+        return VM_ARGS_MAP.get(key);
+      }
+      if (currentIndex >= numArgs) {
+        return null;
+      }
+
+      // Initialize cache if empty
+      if (VM_ARGS_MAP.isEmpty()) {
+        VM_ARGS_MAP = new HashMap<>();
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:41]: this changes the field to be different to the object we originally locked on, so a second thread could now enter this part of the code while we're still mutating things

to simplify this I would make `VM_ARGS_MAP` final with a new `HashMap` - this doesn't take up much space, and since we're doing this under a lock we don't need to use a concurrent map

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 7]
```diff
@@ -9,18 +9,102 @@
...
   private static final List<String> VM_ARGS = findVmArgs();
+  private static Map<String, String> VM_ARGS_MAP = Collections.emptyMap();
+  private static int currentIndex = 0; // Tracks the last processed index in VM_ARGS
 
   public static List<String> getVmArgs() {
     return VM_ARGS;
   }
 
+  public static String getArgValue(String key) {
+    int numArgs = VM_ARGS.size();
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:28]: we could move this down to just before the `while` loop which uses it

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 8]
```diff
@@ -9,18 +9,102 @@
...
+    // Lazy population of cache
+    synchronized (VM_ARGS_MAP) {
+      // Double check after acquiring lock
+      if (VM_ARGS_MAP.containsKey(key)) {
+        return VM_ARGS_MAP.get(key);
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:33]: we don't expect `null` values in the map, so you could avoid the double lookup:
```suggestion
      String value = VM_ARGS_MAP.get(key);
      if (value != null) {
        return value;
```

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 9]
```diff
@@ -9,18 +9,102 @@
...
+      }
+      if (currentIndex >= numArgs) {
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:35]: Once`VM_ARGS_MAP` is made final and the small block of code goes away this check could also be dropped, as the `while` loop does the same check

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 10]
```diff
@@ -9,18 +9,102 @@
...
+      }
+
+      // Process remaining args
+      while (currentIndex < numArgs) {
+        populateCache(VM_ARGS.get(currentIndex));
+        currentIndex++;
+
+        // Check if we found our key
+        if (VM_ARGS_MAP.containsKey(key)) {
+          return VM_ARGS_MAP.get(key);
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:51]: As above, since we know we're not caching `null` values we can avoid the extra lookup:
```suggestion
        value = VM_ARGS_MAP.get(key);
        if (value != null) {
          return value;
```

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 11]
```diff
@@ -9,18 +9,102 @@
...
+        VM_ARGS_MAP.put(argKey, argValue);
+      } else {
+        VM_ARGS_MAP.put(arg, "");
+      }
+    } else {
+      VM_ARGS_MAP.put(arg, "");
+    }
+  }
+
+  public static boolean argExists(String key) {
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:74]: with the above improvements, I would change this to just call `getArgValue(key) != null` because performance-wise the difference between `get` and `containsKey` will be minimal here.

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 12]
```diff
@@ -9,18 +9,102 @@
...
+      }
+
+      // Process remaining args
+      while (currentIndex < numArgs) {
+        String arg = VM_ARGS.get(currentIndex);
+        populateCache(arg);
+        currentIndex++;
+
+        // Check if we found our key
+        if (key.equals(arg)) {
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:95]: wouldn't this miss system properties? i.e. those would be split into a separate key and value in `VM_ARGS`, but here we are just comparing the full `-D...` argument and not the split key

File: components/cli/src/main/java/datadog/cli/CLIHelper.java [hunk 13]
```diff
@@ -9,18 +9,102 @@
...
+        }
+      }
+
+      return false;
+    }
+  }
+
   @SuppressForbidden
   private static List<String> findVmArgs() {
+    List<String> rawArgs;
```
Comment (mcculls) [inline components/cli/src/main/java/datadog/cli/CLIHelper.java:106]: This variable can go away and the `findVmArgs` method reverted to its previous state.

File: components/yaml/build.gradle.kts [hunk 14]
```diff
@@ -11,4 +11,7 @@ jmh {
 // https://repo1.maven.org/maven2/org/yaml/snakeyaml/2.4/snakeyaml-2.4.pom
 dependencies {
   implementation("org.yaml", "snakeyaml", "2.4")
+  implementation(project(":components:cli"))
+  testImplementation(project(":utils:test-utils"))
+  testImplementation(project(":internal-api"))
```
Comment (mcculls) [inline components/yaml/build.gradle.kts:16]: which test update requires `internal-api`?

File: components/yaml/src/main/java/datadog/yaml/YamlParser.java [hunk 15]
```diff
@@ -1,19 +1,97 @@
...
+    }
+  }
+
+  /**
+   * Processes a YAML template by replacing all template variables with their corresponding values.
+   * Template variables are enclosed in double curly braces, e.g. {{variable}}. Returns the
+   * processed content with all template variables resolved.
+   */
+  static String processTemplate(String content) throws IOException {
+    StringBuilder result = new StringBuilder(content.length());
```
Comment (mcculls) [inline components/yaml/src/main/java/datadog/yaml/YamlParser.java:29]: if we check for the non-existence of `{{` first then we can just return `content` and avoid duplicating the string contents.

File: components/yaml/src/main/java/datadog/yaml/YamlParser.java [hunk 16]
```diff
@@ -1,19 +1,97 @@
...
 public class YamlParser {
   // Supports clazz == null for default yaml parsing
   public static <T> T parse(String filePath, Class<T> clazz) throws IOException {
     Yaml yaml = new Yaml();
-    try (FileInputStream fis = new FileInputStream(filePath)) {
-      if (clazz == null) {
-        return yaml.load(fis);
-      } else {
-        return yaml.loadAs(fis, clazz);
+    String content = new String(Files.readAllBytes(Paths.get(filePath)));
```
Comment (mcculls) [inline components/yaml/src/main/java/datadog/yaml/YamlParser.java:13]:
```suggestion
    String content = new String(Files.readAllBytes(Paths.get(filePath)), StandardCharsets.UTF_8);
```
this avoids using `new String(byte[])` which will use the platform encoding, which may not be UTF_8

File: components/cli/src/main/java/datadog/cli/CLIHelper.java (hunk 9 above)
Comment (mtoffl01) [reply components/cli/src/main/java/datadog/cli/CLIHelper.java:35]: great catch

File: components/cli/src/main/java/datadog/cli/CLIHelper.java (hunk 12 above)
Comment (mtoffl01) [reply components/cli/src/main/java/datadog/cli/CLIHelper.java:95]: It's expected that a user will specify just the key in their stable config file - e.g., `process_arguments: [-Dkey]` - which would invoke `argExists` will argument `-Dkey`.

File: components/yaml/build.gradle.kts (hunk 14 above)
Comment (mtoffl01) [reply components/yaml/build.gradle.kts:16]: ah, this was leftover from before I moved test helpers into test-utils

Comment (mtoffl01) [reply components/yaml/build.gradle.kts:16]: The new YamlParserTest.groovy file needs access to StableConfigYaml class, which is in internal-api

File: components/yaml/src/main/java/datadog/yaml/YamlParser.java (hunk 15 above)
Comment (mtoffl01) [reply components/yaml/src/main/java/datadog/yaml/YamlParser.java:29]: `content` is just the entire file contents at this point, which may or may not have 1+ "template variables" enclosed in `{{}}`. How would we do a quick check up top without iterating over the file contents?

Comment (mtoffl01) [reply components/yaml/src/main/java/datadog/yaml/YamlParser.java:29]: Do you mean doing a check up top to see whether the file contents contain `{{` at all? 
Can you explain how that would be more optimal than what we have now under the `while` loop, which is iterating through the file contents and incrementally checking for the existence of `{{` in the rest of the file? I understand it would avoid duplicating the string contents in the case where no template variable exists, but it would also introduce an extra iteration of the file contents when `{{` does exist.

Comment (mcculls) [reply components/yaml/src/main/java/datadog/yaml/YamlParser.java:30]: You can do `int openIndex = content.indexOf("{{")` first - if that returns `-1` then just return `content`, no need for any copying.

Otherwise you enter the while loop and use the existing `openIndex` value, i.e. you can remove the `indexOf` call and check at the top of the while loop. Then at the bottom of the while loop you do:
```
openIndex = content.indexOf("{{", closeIndex + 2);
if (openIndex == -1) {
  result.append(content, closeIndex + 2, content.length);
  break;
}
```

Basically by careful use of indexes you should be able to avoid continually grabbing substrings of `content` and putting them in `rest` (which will cause GC churn) and instead add sections of `content` outside of `{{` and `}}`.

File: components/yaml/build.gradle.kts (hunk 14 above)
Comment (mcculls) [reply components/yaml/build.gradle.kts:16]: I would prefer to have a separate example class, outside of `StableConfigYaml` for testing purposes

Otherwise you introduce a dependency from what is currently a nicely decoupled component to `internal-api`

File: internal-api/src/main/java/datadog/trace/bootstrap/config/provider/StableConfigParser.java [hunk 17]
```diff
@@ -166,14 +172,100 @@ static boolean selectorMatch(String origin, List<String> matches, String operato
...
+    if (templateVar.startsWith(ENVIRONMENT_VARIABLES_PREFIX) && templateVar.endsWith("']")) {
+      String envVar =
+          templateVar
+              .substring(ENVIRONMENT_VARIABLES_PREFIX.length(), templateVar.length() - 2)
+              .trim();
+      if (envVar.isEmpty()) {
+        throw new IOException("Empty environment variable name in template");
+      }
+      String value = System.getenv(envVar.toUpperCase());
+      if (value == null || value.isEmpty()) {
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/bootstrap/config/provider/StableConfigParser.java:246]: One question - will users ever want to substitute a variable with an empty string?

Currently we don't support this, because the variable will be replaced with `UNDEFINED` - but a user might want to have some kind of optional variable that may have a value or optionally be the empty string...

Comment (mtoffl01) [reply internal-api/src/main/java/datadog/trace/bootstrap/config/provider/StableConfigParser.java:246]: Addressing in a separate PR: https://github.com/DataDog/dd-trace-java/pull/8759

File: utils/test-utils/src/main/groovy/datadog/trace/test/util/FileUtils.groovy [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardOpenOption
+
+class FileUtils {
+  // Use this for writing a string directly into a file
+  static writeFileRaw(Path filePath, String data) {
+    StandardOpenOption[] openOpts = [StandardOpenOption.WRITE] as StandardOpenOption[]
+    Files.write(filePath, data.getBytes(), openOpts)
+  }
```
Comment (PerfectSlayer) [inline utils/test-utils/src/main/groovy/datadog/trace/test/util/FileUtils.groovy:12]: Is that useful to create a dedicated method for a single call which can already be written: `Files.write(path, data.getBytes())`?
Default open options looks good enough for the testing, aren't they?
And I would not recommend re-using this method as it hide the encoding / charset to used (which is usually not recommended).

File: utils/test-utils/src/main/groovy/datadog/trace/test/util/FileUtils.groovy [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+  static Path tempFile() {
+    try {
+      return Files.createTempFile("testFile_", ".yaml")
+    } catch (IOException e) {
+      println "Error creating file: ${e.message}"
+      e.printStackTrace()
+      return null
+    }
```
Comment (PerfectSlayer) [inline utils/test-utils/src/main/groovy/datadog/trace/test/util/FileUtils.groovy:21]: Similarly, this is one-liner helper. 
Checking the exception is not required is groovy so you can directly write `Files.createTempFile()` in your test, and the exception handling won't be useful as you will certainly end up with another exception (NullPoienterException) if something fails.

Additionally, Spock and test framework usually provide tools for temporary files (for example `@TempDir` for Spock: https://spockframework.org/spock/docs/2.3/all_in_one.html#_temp_dir)

**Review by PerfectSlayer — COMMENTED** [review]
I would avoid creating one-liner helpers in `internal-api`.
They can be replaced by their equivalent static method calls from JDK.
